### ISSUE TITLE
Using Aeolus with the correct ocaml and atdgen releases

### ISSUE THREAD

**Issue by emnaro** [aeolus-project/metis#2]
Hello,
I am trying to install metis for research purpose.
I am new to ocaml, so my question may be not related to metis.
I tried to run the make all command but the following error is displayed, despite I have already installed atdgen using the apt install command.

`
$ make all
#ocamlbuild -no-hygiene -classic-display -use-ocamlfind metis.native
ocamlbuild -pp 'camlp4o pa_macro.cmo -UVERBOSE' -no-hygiene -classic-display -use-ocamlfind metis.native  ocamlfind ocamlopt -c -package atdgen -package extlib -pp 'camlp4o pa_macro.cmo -UVERBOSE' -I src/datatypes -I src -I src/engine -o src/datatypes/datatypes_j.cmx src/datatypes/datatypes_j.ml + ocamlfind ocamlopt -c -package atdgen -package extlib -pp 'camlp4o pa_macro.cmo -UVERBOSE' -I src/datatypes -I src -I src/engine -o src/datatypes/datatypes_j.cmx src/datatypes/datatypes_j.ml
File "src/datatypes/datatypes_j.ml", line 40, characters 2-23:
Error: Unbound module Ag_oj_run
Command exited with code 2.
Makefile:7: recipe for target 'all' failed
make: *** [all] Error 10
`

The module Ag_oj_run is not bound. 
I have the atdgen package under  /usr/lib/ocaml.
Do I have to do something to tell metis about the location of atdgen or am I missing something?
Thanks.

**Comment by emnaro** [comment]
I am posting un update of the issue.
It turns out that the Aeolus/metis tool uses the old atdgen package.
Here is the old git rep https://github.com/mjambon/atdgen
and here is the new one:
https://github.com/mjambon/atd/tree/master/atdgen
I am trying to use Aeolus with few  modifications. To use the old atdgen, I downoaled the sources, added the rep under my metis copy and modified the following line in the makefile in order to point to the atdgen src files, including the previously missing Ag_oj_run module:
```
RMPPFLAGS := -pp 'camlp4o pa_macro.cmo -UVERBOSE' -I atdgen-master/src/ 
```
I also refer to Yoson package as follows:
```
all:
	#ocamlbuild $(OBFLAGS) metis.native
	ocamlbuild -pkg yojson $(RMPPFLAGS) $(OBFLAGS) metis.native
```
Now, during the make, the atdgen module is recongnized as well as Yojson.
However, I still get the following error:
```
File "atdgen-master/src/ag_oj_run.ml", line 220, characters 43-45:

Error: Unbound type parameter 'a
```

This is the error line togeter with the two following lines:

```
(* We want an identity function that is not inlined *)

type identity_t = { mutable _identity : 'a.'a -> 'a }

let identity_ref = { _identity = (fun x -> x) }

let identity x = identity_ref._identity x
```
The identity is only called  twice, as follows in the  datatypes_j.ml file:

```
with Yojson.End_of_object -> (
        if !bits0 <> 0x3 then Ag_oj_run.missing_fields [| !bits0 |] [| "u_cname"; "u_automaton" |];

        Ag_oj_run.identity x
```
Since  I am a newbie with Ocaml, I wonder what are the  changes that I may do in order to be able to make the project?

**Comment by emnaro** [comment]
I found a solution and finally could compile and execute Aeolus.  I post it in order to make it possible for other users to be aware of  a slight incompatibility of the new  ocaml and atdgen  packages  with the Aeolus source code.
There is some incompatibility between the `identity` from `Ag_oj_run` which is called  in `datatypes_j.ml` in the Aeolus source code and the new version of OCaml. Despite I called `Oj_run` of the new `atdgen` version  instead of `Ag_oj_run` of the old `atdgen,` a build error  persisted. I tried to slightly modify the source code to eliminate this building error in order to be able to make all, but an error was generated during the execution. I don´t know the Ocaml language sufficiently to be sure to correctly modify the source code.
Finally, I though that it may be a simple incompatibility issue and therefore installed the old version of Ocaml together with the old version of `atdgen.` I have chosen releases of 2014 or 2013 such that they are compatible with  Aeolus which was developed  between  5 and 6 years ago. Following these steps, I had no problem with the Aeolus building and running without any change in its source code.
Here are the commands with the correct versions:
`
$opam switch  4.01.0

$eval `opam config env`

$ which ocaml

$ opam install extlib

$ opam install yojson

$opam install "atdgen=1.3.1"

$make clean

$make all
`
